pkg/utils/handler.go

```diff
@@ -79,36 +120,88 @@ func (h *Handler) AddMiddleware() {
 }
 
 func (h *Handler) UpdateAgents(rw http.ResponseWriter, r *http.Request, rp httprouter.Params) {
-	agentData := AgentInfo{}
+	agentData := ext_v1.AgentSpec{}
+
 	if err := ProcessRequest(r, &agentData, rw); err != nil {
 		return
 	}
 
 	agentData.LastUpdated = time.Now()
-	glog.V(10).Infof("Updating the agents cache with value: %v", agentData)
+	glog.V(10).Infof("Updating the agents resource with value: %v", agentData)
+
 	agentName := rp.ByName("name")
-	if _, exists := h.AgentCache[agentName]; !exists {
+	_, err := h.ExtensionsClientset.Agents().Get(agentName)
+
+	if err != nil {
+		glog.V(5).Info(err)
+	}
+
+	if api_errors.IsNotFound(err) {
 		h.Metrics[agentName] = NewAgentMetrics(&agentData)
 		h.cleanCacheOnDemand(nil)
 	}
+
 	UpdateAgentBaseMetrics(h.Metrics[agentName], true, false)
 	UpdateAgentProbeMetrics(agentData, h.Metrics[agentName])
-	h.AgentCache[agentName] = agentData
+
+	agent := &ext_v1.Agent{
+		ObjectMeta: meta_v1.ObjectMeta{
+			Name: agentName,
+		},
+		Spec: agentData,
+		Status: ext_v1.AgentStatus{
+			State:   ext_v1.AgentStateCreated,
+			Message: "Created, not processed yet",
+		},
+	}
+
+	agent, err = h.ExtensionsClientset.Agents().Create(agent)
+
+	if err != nil {
+		glog.V(5).Info(err)
+	}
+
+	err = ext_client.WaitForAgentInstanceProcessed(h.ExtensionsClientset, agentName)
```

Comment:
I'm not sure it is a good idea to wait synchronously for such stuff. Seems, we need queue here and in onAdd().